packages/compiler/src/core/checker.ts

```diff
@@ -2778,6 +2790,164 @@ export function createChecker(program: Program): Checker {
     return (resolved?.declarations.filter((n) => isTemplatedNode(n)) ?? []) as TemplateableNode[];
   }
 
+  function getReferencedModel(
+    propertyNode: ObjectLiteralPropertyNode | ModelPropertyNode
+  ): Model | undefined {
+    const ARRAY_MODEL_FLAG = "@@array_model_flag@@";
+
+    if (propertyNode.kind === SyntaxKind.ModelProperty) {
+      return getReferencedModelFromDecoratorArgument(propertyNode);
+    } else if (propertyNode.kind === SyntaxKind.ObjectLiteralProperty) {
+      let r = getReferencedModelFromScalarConstructor(propertyNode);
+      r ??= getReferencedModelFromConstAssignment(propertyNode);
+      r ??= getReferencedModelFromDecoratorArgument(propertyNode);
+      return r;
+    } else {
+      return undefined;
+    }
+
+    function pushToModelPath(node: Node, path: string[]) {
+      if (node.kind === SyntaxKind.ArrayLiteral || node.kind === SyntaxKind.TupleExpression) {
+        path.unshift(ARRAY_MODEL_FLAG);
```

Comment:
I think you should keep the index here instead of this string, otherwise cases like this won't work

```
const [Foo, Bar] = #[{ ... auto complete foo }, ]
```

Reply:
do you mean an array can have different type? could you help to describe more about the scenario? the sample you gave above seems not valid as below
![image](https://github.com/microsoft/typespec/assets/6038235/78107920-594b-4f86-9681-3bf742178990)

Reply:
yes sorry missed the const name
```
model Foo {
  foo: string;
}
model Bar {
  bar: string;
}


const vals: [Foo, Bar] = #[#{foo: "abc"}, #{bar: "abc"}]
```
https://cadlplayground.z22.web.core.windows.net/prs/3280/?c=bW9kZWwgRm9vIHsKICBmb286IHN0cmluZzsKfQrGHUJhcsUdYmFyzB0KCmNvbnN0IHZhbHM6IFtGb28sxCtdID0gI1sje8VLImFiYyJ9LCAje8U9xg9d&e=%40typespec%2Fopenapi3&options=%7B%7D

Reply:
technically need different path if its just random index or a specific one so might need to keep that special key, though can you replace it with a `symbol()`

Reply:
By "random index or a specific one", do you mean Tuple and Array<T>. Anyway, i have updated the code to support Tuple, Array<T> and model. Please help to check and let me know if there are other cases I missed, thanks.